### PR TITLE
Fix SourceText.GetChanges exception on merging certain changes

### LINKED ISSUE #22289: ArgumentOutOfRangeException in CompletionService.GetCompletionsAsync
**Version Used**: 
2.3.2

**Steps to Reproduce**:
Unfortunately I don't know how to isolate this. This is something I see in logs of my Roslyn-based project, and I wasn't able to reproduce it locally (however I consistently see it in logs).

I hope you might be able to guess something from the stack trace, or suggest a way to get more information about this.

Stack trace (cleaned up from Application Insights):
```
System.AggregateException: One or more errors occurred.
Microsoft.CodeAnalysis.DocumentState+<GetSyntaxTreeAsync>d__41.MoveNext
Microsoft.CodeAnalysis.CSharp.Completion.Providers.AttributeNamedParameterCompletionProvider+<ProvideCompletionsAsync>d__5.MoveNext
Microsoft.CodeAnalysis.Completion.CompletionServiceWithProviders+<GetContextAsync>d__31.MoveNext
Microsoft.CodeAnalysis.Completion.CompletionServiceWithProviders+<ComputeNonEmptyCompletionContextsAsync>d__24.MoveNext
Microsoft.CodeAnalysis.Completion.CompletionServiceWithProviders+<GetCompletionsAsync>d__22.MoveNext

System.AggregateException: One or more errors occurred.
Microsoft.CodeAnalysis.DocumentState+<IncrementallyParseTreeAsync>d__15.MoveNext

System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
 Parameter name: start
Microsoft.CodeAnalysis.Text.TextSpan..ctor
Microsoft.CodeAnalysis.Text.ChangedText.Merge
Microsoft.CodeAnalysis.Text.ChangedText.Merge
Microsoft.CodeAnalysis.Text.ChangedText.GetChangeRanges
Microsoft.CodeAnalysis.CSharp.CSharpSyntaxTree.WithChangedText
Microsoft.CodeAnalysis.DocumentState.IncrementallyParse
Microsoft.CodeAnalysis.DocumentState+<IncrementallyParseTreeAsync>d__15.MoveNext
```

**Expected Behavior**:
There are no exceptions from `CompletionService`.

### LINKED ISSUE #26305: ChangedText.GetTextChanges should not assert
```csharp
[Fact]
        public void ChangedTextTests()
        {
            var text0 = SourceText.From(@"using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ConsoleApp22
{
    class Class1
    {
        void Method()
        {
            Console

     
        }

    }
}");

            var text1 = text0.WithChanges(new TextChange(TextSpan.FromBounds(221, 228), ""));
            var text2 = text1.WithChanges(new TextChange(TextSpan.FromBounds(228, 229), ""));
            var text3 = text2.WithChanges(new TextChange(TextSpan.FromBounds(228, 228), "."));
            var text4 = text3.WithChanges(new TextChange(TextSpan.FromBounds(229, 229), "w"));
            var text5 = text4.WithChanges(new TextChange(TextSpan.FromBounds(230, 230), "R"));
            var text6 = text5.WithChanges(new TextChange(TextSpan.FromBounds(231, 231), "I"));
            var text7 = text6.WithChanges(new TextChange(TextSpan.FromBounds(232, 232), "T"));
            var text8 = text7.WithChanges(new TextChange(TextSpan.FromBounds(233, 233), "E"));
            var text9 = text7.WithChanges(new TextChange(TextSpan.FromBounds(231, 232), ""));

            var changes = text9.GetTextChanges(text1).ToList();
        }
```

add this test and run. then, this will assert
http://source.roslyn.io/#Microsoft.CodeAnalysis/Text/ChangedText.cs,375

this should return full file content as changed if it can't figure out text changes between 2 given SourceText.

our SourceText doesn't have concept of before or after text like editor's ITextImage (there is no prevous or next pointer in SourceText or versioning). ChangedText implementation of SourceText internally seems have that info, but that shouldn't leak out to outside.

### PR DESCRIPTION
**Pull request by ashmind**
Closes #22289
Closes #26305

See #22289 for full details on the bug.

#### Summary
After certain `WithChanges` calls, `SourceText.GetChanges` is no longer able to handle the `SourceText` state and throws `ArgumentOutOfRangeException` when called.

#### Notes
- Three of the new tests are failing without my change, other tests added just in case.
- I didn't follow "Hello World" pattern in tests, sorry. I found it a bit hard to remember the offsets when using actual words.
- Not fully confident in the change even though the tests pass. The flow is quite complicated and I feel the current coverage might be insufficient.

### CODE REVIEW

File: src/Compilers/Core/Portable/Text/ChangedText.cs [hunk 1]
```diff
@@ -313,11 +313,11 @@ private static ImmutableArray<TextChangeRange> Merge(ImmutableArray<TextChangeRa
...
+                        else if (newChange.Span.Length <= oldChange.NewLength)
                         {
-                            // new change is just an insertion, go ahead and tack it on with old change
-                            AddRange(list, new TextChangeRange(oldChange.Span, oldChange.NewLength + newChange.NewLength));
-                            oldDelta = oldDelta - oldChange.Span.Length + oldChange.NewLength;
+                            // new change is smaller than old, go ahead and tack it on with old change
+                            AddRange(list, new TextChangeRange(oldChange.Span, oldChange.NewLength + newChange.NewLength - newChange.Span.Length));
+                            oldDelta = oldDelta - oldChange.Span.Length + oldChange.NewLength - newChange.Span.Length;
                             oldIndex++;
                             newIndex++;
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Text/ChangedText.cs]: not certain why this is safe/ok.  if newChange and oldChange have different contents, then what's going on here?

Comment (ashmind) [reply src/Compilers/Core/Portable/Text/ChangedText.cs]: It actually merges the list of `TextChangeRange` objects, not `TextChange` objects.
Those don't have contents, just old length and new length, so there is no difference in contents.

The ranges are mapped to the final text later, in `GetTextChanges`.

File: src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs [hunk 2]
```diff
@@ -636,6 +636,32 @@ public void TestMergeChanges_Overlapping_NewInsideOld()
...
+            var original = SourceText.From("01234");
+            var change1 = original.WithChanges(new TextChange(new TextSpan(1, 3), "aa"));
+            var change2 = change1.WithChanges(new TextChange(new TextSpan(2, 0), "bb"));
+
+            var changes = change2.GetTextChanges(original);
+            Assert.Equal("0aa4", change1.ToString());
+            Assert.Equal("0abba4", change2.ToString());
+            Assert.Equal(new[] { new TextChange(new TextSpan(1, 3), "abba") }, changes);
+        }
+
```
Comment (RikkiGibson) [inline src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs]:
```suggestion

        [Fact]
        public void TestMergeChanges_Overlapping_NewInsideOld_AndOldHasLeadingDeletion_SmallerThanLeadingInsertion()
        {
            var original = SourceText.From("012");
            var change1 = original.WithChanges(new TextChange(new TextSpan(1, 1), "aaa"));
            var change2 = change1.WithChanges(new TextChange(new TextSpan(3, 0), "bb"));

            var changes = change2.GetTextChanges(original);
            Assert.Equal("0aaa2", change1.ToString());
            Assert.Equal("0aabba2", change2.ToString());
            Assert.Equal(new[] { new TextChange(new TextSpan(1, 1), "aabba") }, changes);
        }

```

File: src/Compilers/Core/Portable/Text/ChangedText.cs [hunk 3]
```diff
@@ -296,10 +296,10 @@ private static ImmutableArray<TextChangeRange> Merge(ImmutableArray<TextChangeRa
                         // new change starts after old change, but overlaps
                         // add as much of the old change as possible and try again
                         var oldChangeLeadingInsertion = newChange.Span.Start - (oldChange.Span.Start + oldDelta);
-                        AddRange(list, new TextChangeRange(oldChange.Span, oldChangeLeadingInsertion));
-                        oldDelta = oldDelta - oldChange.Span.Length + oldChangeLeadingInsertion;
-                        oldChange = new TextChangeRange(new TextSpan(oldChange.Span.Start, 0), oldChange.NewLength - oldChangeLeadingInsertion);
-                        newChange = new TextChangeRange(new TextSpan(oldChange.Span.Start + oldDelta, newChange.Span.Length), newChange.NewLength);
+                        var oldChangeLeadingDeletion = Math.Min(oldChange.Span.Length, oldChangeLeadingInsertion);
```
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/Text/ChangedText.cs]: OK, I'm talking myself through why this makes sense.

In this line we are figuring out how many characters of the old change occur before the new change. We assume that if the old change deletes characters then it will delete from the left *at most* as many as it is inserting on the left. When the old change's deletion is longer than the old change's leading insertion, the rest of its deletion has to be shifted over to remove the characters after the new change's insertion.

File: src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs [hunk 4]
```diff
@@ -678,6 +704,84 @@ public void TestMergeChanges_Overlapping_OldBeforeNew()
...
+        {
+            var original = SourceText.From("01234");
+            var change1 = original.WithChanges(new TextChange(new TextSpan(1, 3), "aa"));
+            var change2 = change1.WithChanges(new TextChange(new TextSpan(1, 1), "bb"));
+
+            var changes = change2.GetTextChanges(original);
+            Assert.Equal("0aa4", change1.ToString());
+            Assert.Equal("0bba4", change2.ToString());
+            Assert.Equal(new[] { new TextChange(new TextSpan(1, 3), "bba") }, changes);
+        }
```
Comment (RikkiGibson) [inline src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs]: Are these all cases that would have thrown before your change?

BTW, this new form for the tests is significantly easier to reason about, thanks.

File: src/Compilers/Core/Portable/Text/ChangedText.cs [hunk 5]
```diff
@@ -313,11 +313,11 @@ private static ImmutableArray<TextChangeRange> Merge(ImmutableArray<TextChangeRa
                             oldIndex++;
                             goto nextOldChange;
                         }
-                        else if (newChange.Span.Length == 0)
+                        else if (newChange.Span.Length <= oldChange.NewLength)
                         {
-                            // new change is just an insertion, go ahead and tack it on with old change
-                            AddRange(list, new TextChangeRange(oldChange.Span, oldChange.NewLength + newChange.NewLength));
-                            oldDelta = oldDelta - oldChange.Span.Length + oldChange.NewLength;
+                            // new change is smaller than old, go ahead and tack it on with old change
```
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/Text/ChangedText.cs]:
```suggestion
                            // new change deletes fewer characters than old change inserted
                            // add new change insertion, then the remaining trailing characters of the old change insertion
```

File: src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs (hunk 2 above)
Comment (ashmind) [reply src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs]: Thanks, applied!

File: src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs (hunk 4 above)
Comment (ashmind) [reply src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs]: Only 3 of them would throw, but I no longer remember which ones. `TestMergeChanges_Overlapping_NewInsideOld_AndOldHasDeletion` for sure. We can revert the class itself on local and run tests to see.

Comment (RikkiGibson) [reply src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs]: I ended up pulling down the changes after I commented to see what would break, and it was pretty much as you describe :)

I actually needed to pull down the changes and run the whole test class with the debugger with breakpoints on areas you changed to be sure I understood what the code was doing.

Comment (ashmind) [reply src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs]: Thanks for your time! Yep it was a bit hard to follow for me as well, existing tests were extremely useful.

**Comment by RikkiGibson** [comment]
Could we please get a senior review on this? @dotnet/roslyn-compiler

File: src/Compilers/Core/Portable/Text/ChangedText.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/Core/Portable/Text/ChangedText.cs]: I don't know if this will clarify things or not, but here's the way I understand this method:

- We have an original text `t1` and call `t1.WithTextChanges(oldChanges)`. This produces a ChangedText `t2`, storing the `oldChanges` and `t1` in fields.
- When we call `t2.WithTextChanges(newChanges)` it produces a new ChangedText `t3`, storing the `newChanges` and `t2`.
- When we call `t3.GetTextChanges(t1)`, we have to rewrite the `newChanges` to make them occur w.r.t. `t1` instead of w.r.t. `t2`:
  - we have to split individual changes apart when old and new changes overlap with each other
  - we merge adjacent changes together as we go to simplify them.

In the end we get a `TextChangeRange[]` which describes the difference between the original text and the final text.

**Comment by RikkiGibson** [comment]
Ping @dotnet/roslyn-compiler

File: src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs [hunk 6]
```diff
@@ -636,6 +636,45 @@ public void TestMergeChanges_Overlapping_NewInsideOld()
             Assert.Equal("Cool ", changes[0].NewText);
         }
 
+        [Fact]
```
Comment (cston) [inline src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs]: >Fact [](start = 9, length = 4)

Please add a `WorkItem` attribute to each of these three tests and the last test. 
```
[WorkItem(22289, "[https://github.com/dotnet/roslyn/issues/22289](https://github.com/dotnet/roslyn/issues/22289)")]
```

Comment (RikkiGibson) [reply src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs]: Looks like the link might have been auto-formatted, did you mean something like: 

```cs
[WorkItem(22289, "https://github.com/dotnet/roslyn/issues/22289")]
```

File: src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs [hunk 7]
```diff
@@ -678,6 +717,84 @@ public void TestMergeChanges_Overlapping_OldBeforeNew()
...
+        {
+            var original = SourceText.From("01234");
+            var change1 = original.WithChanges(new TextChange(new TextSpan(1, 3), "aa"));
+            var change2 = change1.WithChanges(new TextChange(new TextSpan(1, 1), "bb"));
+
+            var changes = change2.GetTextChanges(original);
+            Assert.Equal("0aa4", change1.ToString());
+            Assert.Equal("0bba4", change2.ToString());
+            Assert.Equal(new[] { new TextChange(new TextSpan(1, 3), "bba") }, changes);
+        }
```
Comment (cston) [inline src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs]: >} [](start = 8, length = 1)

Do we also need to handle the case where the new deletion is larger than the old? It looks like the following fails with or without this PR:
```C#
var original = SourceText.From("01234");
var change1 = original.WithChanges(new TextChange(new TextSpan(1, 3), "aa"));
var change2 = change1.WithChanges(new TextChange(new TextSpan(1, 3), "bb"));

var changes = change2.GetTextChanges(original);
Assert.Equal("0aa4", change1.ToString());
Assert.Equal("0bb", change2.ToString());
```
We can log a bug for this case and handle it separately.

Comment (ashmind) [reply src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs]: Thanks! I'll take a look in a few moments -- if it's easy to fix then maybe just fix it here, if hard (needs a lot of review) then I can do another PR.

File: src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs (hunk 6 above)
Comment (ashmind) [reply src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs]: Done.

File: src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs (hunk 7 above)
Comment (ashmind) [reply src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs]: Seems similar to what I am already doing in `newChange.Span.Start > oldChange.Span.Start + oldDelta`, but that means it's complicated enough for it's own review.

So I would prefer to do as a separate bug/PR just to make review easier (also maybe there is some reuse that can be done between those two cases).

**Comment by RikkiGibson** [comment]
Thanks so much for your contribution @ashmind!
